### DISCUSSION TITLE
Feature Request: gomod support for deprecated modules

### DISCUSSION THREAD

**Discussion by chey** [renovatebot/renovate#27474] [Suggest an Idea]
### Tell us more.

Here is an example module that is deprecated with a suggested alternative: https://pkg.go.dev/github.com/streadway/amqp

I would like to see renovate create a PR for the module alternative.

As a bonus, once renovate makes this change in go.mod, if a user has `gomodUpdateImportPaths` postUpdate option configured, the paths in the code would be corrected as well.

**Comment by rarkins** [comment]
Is this deprecation field/data available as official metadata eg through the GOPROXY result? Or would it require manual crowdsourced provisioning of these ?

**Reply by chey** [reply]
This information can be gathered from the go.mod file.

Example:
`curl https://proxy.golang.org/github.com/streadway/amqp/@v/v1.1.0.mod`

Output:
```
// Deprecated: Consider using the github.com/rabbitmq/amqp091-go package instead.
module github.com/streadway/amqp

go 1.10
```

Once you have that information you'll need to parse the `Deprecated:` message. Take a look at the [grammer](https://go.dev/ref/mod#go-mod-file-grammar) section of the go mod reference.

My suggestion would be to parse that for an alternative module which may or may not be there. Either look for the word "use|using" and/or "instead" combined with a URL that matches a valid go module.

**Reply by rarkins** [reply]
It seems that `// Deprecated:` is formalized but the contents after it is not. It's not possible to determine if a suggested replacement is a 1:1 drop-in replacement or has a completely different API

**Comment by rarkins** [comment]
Converted to https://github.com/renovatebot/renovate/issues/27478

Automatic migration does not seem possible reliably, so we'd just be warning of the deprecation

**Reply by chey** [reply]
This would be a great start. Thank you.
